query/src/main/scala/filodb/query/exec/ApplySubqueryRangeFunction.scala

```diff
@@ -0,0 +1,155 @@
+  subqueryRangeFunction: InternalRangeFunction
+) extends RangeVectorTransformer {
+  override def funcParams: Seq[FuncArgs] = Nil
+
+  override protected[exec] def args: String =
+    s"subqueryEprRangeParams=$subqueryExpRangeParams," +
+    s"subqueryRangeFunction=$subqueryRangeFunction," +
+    s"underlyingLookbackMs=$underlyingLookbackMs,underlyingStepMs=$underlyingStepMs"
+
+  def apply(
+    source: Observable[RangeVector],
+    querySession: QuerySession,
+    limit: Int,
+    sourceSchema: ResultSchema,
+    paramsResponse: Seq[Observable[ScalarRangeVector]]
+  ): Observable[RangeVector] = {
+
+    //val underlyingRangeStepMs = underlyingPeriodicSeriesRange.stepSecs * 1000
+
+    // Need to skip these many samples for window operations
+    //val skip = (subqueryStepMs / underlyingRangeStepMs).toInt
+    val valColType = ResultSchema.valueColumnType(sourceSchema)
+    val rangeFuncGen = RangeFunction.generatorFor(
+      sourceSchema,
+      Some(subqueryRangeFunction),
+      valColType,
+      querySession.queryConfig, funcParams, false
+    )
+    source.map { rv =>
+      IteratorBackedRangeVector(
+        rv.key,
+        new SubquerySlidingWindowIterator(
+          subqueryExpRangeParams.startSecs * 1000,
+          subqueryExpRangeParams.endSecs * 1000,
+          subqueryExpRangeParams.stepSecs * 1000,
+          underlyingLookbackMs,
+          underlyingStepMs,
+          rv.rows,
+          rangeFuncGen,
+          querySession.queryConfig
+        ),
+        None
+      )
+    }
+  }
+
+  override def schema(source: ResultSchema): ResultSchema = {
+    // since range function is NOT optional, the resulting vectors length will be
+    // of size 1, not the length of the underlying PeriodicSeries
+    source.copy(fixedVectorLen = Some(1))
+  }
+
+}
+
+// sum_over_time(foo[5m:1m])[3m:90s]
+class SubquerySlidingWindowIterator(
+   startMs: Long, // now - 3m
+   endMs : Long, // now
+   stepMs: Long, // 90s
+   underlyingLookbackMs: Long, //5m
+   underlyingStepMs: Long, //1m
+   rvc: RangeVectorCursor,
+   rfGenerator: RangeFunctionGenerator, //TODO require reset for all range functions
+   //rangeFunction: RangeFunction,
+   queryConfig: QueryConfig
+ ) extends RangeVectorCursor {
+  private val sampleToEmit = new TransientRow()
+
+  // samples queue that will have ALL of the samples in the current range
+  // whether they are applicable for the current window or not
+  private val rangeQueue = new IndexedArrayQueue[TransientRow]()
+
+  // to avoid creation of object per sample, we use a pool
+  val windowSamplesPool = new TransientRowPool()
+
+  private var ts = 0L
+  private var curWindowStartMs = startMs - underlyingLookbackMs
+  private var curWindowEndMs = startMs
+  private var finished = false
+
+  override def close(): Unit = rvc.close()
+
+  override def hasNext: Boolean = {
+    if (finished) {
+      false
+    } else {
+      rvc.hasNext
+    }
+  }
+
+  override def next(): TransientRow = {
```

Comment:
Please document overall strategy here - it is not clear.
What is the expected runtime complexity here ? Is it O(samples) or is it O(samples * windowLength). I ask because we seem to be clearing out buffer each time.

Reply:
documented

Reply:
removed the entire class